File: java/src/main/java/org/samples/todos/repository/TaskRepository.java
package org.samples.todos.repository;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.samples.todos.model.TaskGroup;

import java.io.File;
import java.io.IOException;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;

public class TaskRepository {

    private final String taskFileName;
    private final ObjectMapper mapper;

    public TaskRepository(final String taskFileName, final ObjectMapper mapper) {
        this.taskFileName = taskFileName;
        this.mapper = mapper;
    }

    public List<TaskGroup> load() {
        List<TaskGroup> taskGroups = new ArrayList<>();

        try {
            taskGroups = mapper.readValue(Paths.get(taskFileName).toFile(), new TypeReference<>() {
            });
        } catch (IOException e) {
            System.err.printf(String.format("Failed to load tasks from file: %s%n", taskFileName));
            e.printStackTrace();
        }

        return taskGroups;
    }

    public void save(final List<TaskGroup> taskGroups) {
        try {
            mapper.writeValue(new File(taskFileName), taskGroups);
        } catch (IOException e) {
            System.err.printf(String.format("Failed to save tasks to file: %s", taskFileName));
            e.printStackTrace();
        }
    }
}

File: src/main/kotlin/org/samples/todos/repository/TaskRepository.kt
package org.samples.todos.repository

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import org.samples.todos.model.TaskGroup
import java.io.File
import java.io.IOException
import java.nio.file.Paths


class TaskRepository(private val taskFileName: String, private val mapper: ObjectMapper) {

    fun load(): MutableList<TaskGroup> {
        try {
            return mapper.readValue(Paths.get(taskFileName).toFile())
        } catch (e: IOException) {
            System.err.printf(String.format("Failed to load tasks from file: %s%n", taskFileName));
            e.printStackTrace();
            return mutableListOf()
        }
    }

    fun save(taskGroups: MutableList<TaskGroup>) {
        try {
            mapper.writeValue(File(taskFileName), taskGroups)
        } catch (e: IOException) {
            System.err.printf(String.format("Failed to save tasks to file: %s", taskFileName))
            e.printStackTrace()
        }
    }
}

